File: Interview-Preparation/MissionJune2020/src/com/interviewbit/LinkedLists/ReverseLinkedList.java
package com.interviewbit.LinkedLists;

public class ReverseLinkedList {

    static class ListNode {
        int val;
        ListNode next;
        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public static ListNode reverseList(ListNode head) {
        ListNode curr = head, prev = null, next = null;
        while (curr != null) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }

    public static void printList(final ListNode head) {
        ListNode curr = head;
        while (curr != null) {
            System.out.print(curr.val + " ");
            curr = curr.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);
        head.next.next.next.next = new ListNode(5);
        printList(head);

        printList(reverseList(head));

    }
}

File: InterviewBit/20210701_2204/L4-Linked_Lists/reverse-linked-list.cpp
#include <iostream>
using namespace std;

struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

ListNode* reverse(ListNode *head) {
    if (!head) return head;
    ListNode *next = reverse(head->next);
    head->next->next = head;
    head->next = nullptr;
    return next;
}

ListNode *reverseList(ListNode *A) {
  ListNode *curr = A, *next = nullptr, *prev = nullptr;
  while (curr) {
    next = curr->next;
    curr->next = prev;
    prev = curr;
    curr = next;
  }
  return prev;
}

int main(int argc, char const *argv[]) {
  /* code */
  return 0;
}
